pkg/controller/volume/attachdetach/reconciler/reconciler.go

```diff
@@ -251,6 +251,9 @@ func (rc *reconciler) reconcile() {
 			if err != nil {
 				// Skip detaching this volume if unable to update node status
 				klog.ErrorS(err, "UpdateNodeStatusForNode failed while attempting to report volume as attached", "volume", attachedVolume)
+				// Add volume back to ReportAsAttached if UpdateNodeStatusForNode call failed so that node status updater will add it back to VolumeAttached list.
+				// It is needed here too because DetachVolume is not call actually and we keep the data consistency for every reconcile.
+				rc.actualStateOfWorld.AddVolumeToReportAsAttached(attachedVolume.VolumeName, attachedVolume.NodeName)
```

Comment:
The change looks good, the only small thing is that once calling AddVolumeToReportAsAttached, the statusUpdateNeeded will be set true, and node status will update again, but actually nothing has changed.

I think in the future we might rethink how to simply logic related update volume attached list, the current logic is proven to be error-prune.

Reply:
It seems to me the problem is - `UpdateNodeStatusForNode` fails but `UpdateNodeStatuses` succeeds later on then, second time a detach is performed then `RemoveVolumeFromReportAsAttached` will fail because volume was previously removed from the list of `volumesToReportAsAttached` volumes. 

I wonder if a better fix here would be to return a different error from function `RemoveVolumeFromReportAsAttached` and if volume is already removed from list of `reportAsAttached` volumes then let the detach proceed. 

This would avoid the flapping of states that @jingxu97 talking about.

Reply:
@gnufied I think what you said is another problem again, this problem should be as follows step:
1. Pod is deleting
2. remove volume from reportedAsAttached, so node status updater will
update volumeAttached list
3. `UpdateNodeStatusForNode ` fails but `UpdateNodeStatuses ` succeeds later on then
4. another pod is added which using the same volume on the same node
5. reconciler loops and it will NOT try to tigger detach anymore because asw is consistent with dsw

At this point, volume is still attached and in actual state, but
volumeAttached list in node status does not has this volume anymore, and
will block volume mount from kubelet.

Reply:
Hmm yeah I think this patch however does fix that error too. I agree with @jingxu97 that this logic is too error prone and should be simplified.